### PR TITLE
EOS: Add a JSON option when running a cli command

### PR DESCRIPTION
**Pull request by dasturiasArista**
pyeapi supports JSON as an option for the cli commands. However right
now the JSON option isn't exposed to users of this API. This exposes the
JSON option as well.

I manually tested this with an Arista switch and verified the changes work.

### CODE REVIEW

**Comment by ktbyers** [comment]
Tests are failing on this change.

**Comment by dasturiasArista** [comment]
> Tests are failing on this change.

I was noticing that. I'm wondering the best way to handle adding functionality would be. It looks like tests are failing because I've added an extra argument to this function that it's not expecting. Would it ok to modify the test to allow the extra args in the EOS case?

**Comment by ktbyers** [comment]
So the test issue is that all of the NAPALM drivers (junos, eos, ios, etc) must have the same signature for the method:

```
    def cli(self, commands, encoding="text"):
```

So if we add the `encoding` argument, then we have to add it for all of the drivers.

You can always access pyeapi via napalm using:

```
napalm_object.device
```

I guess I wonder if we really need to be able to execute the `cli` method but with JSON output. Note, I basically never use the `cli` method...so probably would be better if @mirceaulinic or other users of CLI commented on this and whether it is worth doing this.

**Comment by dasturiasArista** [comment]
> So the test issue is that all of the NAPALM drivers (junos, eos, ios, etc) must have the same signature for the method:
> 
> ```
>     def cli(self, commands, encoding="text"):
> ```
> 
> So if we add the `encoding` argument, then we have to add it for all of the drivers.
> 
> You can always access pyeapi via napalm using:
> 
> ```
> napalm_object.device
> ```
> 
> I guess I wonder if we really need to be able to execute the `cli` method but with JSON output. Note, I basically never use the `cli` method...so probably would be better if @mirceaulinic or other users of CLI commented on this and whether it is worth doing this.

Ok. that seems reasonable. I was unaware that the "device" attribute was considered a stable part of the api. I have proposed this solution for the customer trying to use this to confirm if this works for them or not.

**Comment by ktbyers** [comment]
Yeah, I would think we would `not` change `napalm_object.device` (at least not without really good reason and without following semantic versioning i.e. a major release) as that would break lots of things.

It is possible this `napalm_object.device` convention is not followed in napalm-community drivers, but it should be followed in napalm core.

**Comment by ktbyers** [comment]
@dasturiasArista Yeah, let us know what you find and if that solution works.

**Comment by dasturiasArista** [comment]
> @dasturiasArista Yeah, let us know what you find and if that solution works.

Yes will do! Not to keep you hanging, still waiting from the response from the customer.

**Comment by mirceaulinic** [comment]
Just to add to Kirk's comment further:

> So if we add the encoding argument, then we have to add it for all of the drivers.

We have a similar request for Junos (#1500) so it'd be great if you could go ahead with this and I can take care of the Junos implementation.

**Comment by dasturiasArista** [comment]
> Just to add to Kirk's comment further:
> 
> > So if we add the encoding argument, then we have to add it for all of the drivers.
> 
> We have a similar request for Junos (#1500) so it'd be great if you could go ahead with this and I can take care of the Junos implementation.

Ok. I can go ahead and add the argument to all of the drivers, but for those that don't support that encoding i'll just have it raise an error. How do you feel about raising a `ValueError`?

**Comment by dasturiasArista** [comment]
> > Just to add to Kirk's comment further:
> > > So if we add the encoding argument, then we have to add it for all of the drivers.
> > 
> > 
> > We have a similar request for Junos (#1500) so it'd be great if you could go ahead with this and I can take care of the Junos implementation.
> 
> Ok. I can go ahead and add the argument to all of the drivers, but for those that don't support that encoding i'll just have it raise an error. How do you feel about raising a `ValueError`?

Btw I have proactively made the changes we talked about here. Also, each method will check that it's valid and throw a `ValueError` if it's not a valid encoding.

File: napalm/eos/eos.py [hunk 1]
```diff
@@ -885,17 +885,21 @@ def get_lldp_neighbors_detail(self, interface=""):
                 )
         return lldp_neighbors_out
 
-    def cli(self, commands):
+    def cli(self, commands, encoding="text"):
+        if encoding not in ("text", "json"):
+            raise ValueError("%s is not a supported encoding" % encoding)
```
Comment (mirceaulinic) [inline napalm/eos/eos.py:890]:
```suggestion
            raise NotImplementedError("%s is not a supported encoding" % encoding)
```

File: napalm/ios/ios.py [hunk 2]
```diff
@@ -2375,6 +2375,8 @@ def cli(self, commands):
             'show clock': u'*22:01:51.165 UTC Thu Feb 18 2016'}
 
         """
+        if encoding not in ("text",):
+            raise ValueError("%s is not a supported encoding" % encoding)
```
Comment (mirceaulinic) [inline napalm/ios/ios.py:2379]:
```suggestion
            raise NotImplementedError("%s is not a supported encoding" % encoding)
```

File: napalm/iosxr/iosxr.py [hunk 3]
```diff
@@ -885,7 +885,9 @@ def get_lldp_neighbors_detail(self, interface=""):
 
         return lldp_neighbors
 
-    def cli(self, commands):
+    def cli(self, commands, encoding="text"):
+        if encoding not in ("text",):
+            raise ValueError("%s is not a supported encoding" % encoding)
```
Comment (mirceaulinic) [inline napalm/iosxr/iosxr.py:890]:
```suggestion
            raise NotImplementedError("%s is not a supported encoding" % encoding)
```

File: napalm/junos/junos.py [hunk 4]
```diff
@@ -1012,8 +1012,10 @@ def get_lldp_neighbors_detail(self, interface=""):
 
         return lldp_neighbors
 
-    def cli(self, commands):
+    def cli(self, commands, encoding="text"):
         """Execute raw CLI commands and returns their output."""
+        if encoding not in ("text",):
+            raise ValueError("%s is not a supported encoding" % encoding)
```
Comment (mirceaulinic) [inline napalm/junos/junos.py:1018]:
```suggestion
            raise NotImplementedError("%s is not a supported encoding" % encoding)
```

File: napalm/nxos/nxos.py [hunk 5]
```diff
@@ -1127,7 +1127,11 @@ def get_bgp_neighbors(self) -> Dict[str, models.BGPStateNeighborsPerVRFDict]:
             results[vrf_name] = result_vrf_dict
         return results
 
-    def cli(self, commands: List[str]) -> Dict[str, Union[str, Dict[str, Any]]]:
+    def cli(
+        self, commands: List[str], encoding: str = "text"
+    ) -> Dict[str, Union[str, Dict[str, Any]]]:
+        if encoding not in ("text",):
+            raise ValueError("%s is not a supported encoding" % encoding)
```
Comment (mirceaulinic) [inline napalm/nxos/nxos.py:1134]:
```suggestion
            raise NotImplementedError("%s is not a supported encoding" % encoding)
```

File: napalm/nxos_ssh/nxos_ssh.py [hunk 6]
```diff
@@ -787,7 +787,9 @@ def get_bgp_neighbors(self):
         # FIX -- need to merge IPv6 and IPv4 AFI for same neighbor
         return bgp_dict
 
-    def cli(self, commands):
+    def cli(self, commands, encoding="text"):
+        if encoding not in ("text",):
+            raise ValueError("%s is not a supported encoding" % encoding)
```
Comment (mirceaulinic) [inline napalm/nxos_ssh/nxos_ssh.py:792]:
```suggestion
            raise NotImplementedError("%s is not a supported encoding" % encoding)
```

**Review by mirceaulinic — COMMENTED** [review]
Looks good to me, just requesting to raise `NotImplementedError` instead of `ValueError`.

**Review by mirceaulinic — APPROVED** [review]
Thanks @dasturiasArista!

**Pull request merged**